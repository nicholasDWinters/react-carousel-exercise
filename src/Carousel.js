import React, { useState } from "react";
import "./Carousel.css";
import image1 from "./image1.jpg";
import image2 from "./image2.jpg";
import image3 from "./image3.jpg";
import Card from "./Card";

function Carousel(props) {
  const [cardIdx, setCardIdx] = useState(0);
  const card = props.cardData[cardIdx];
  const total = props.cardData.length;
  const goForward = () => setCardIdx(cardIdx + 1);
  const goBackward = () => setCardIdx(cardIdx - 1);
  let rightHide = cardIdx === 2 ? 'Carousel-hidden fas fa-chevron-circle-right fa-2x' : 'fas fa-chevron-circle-right fa-2x';
  let leftHide = cardIdx === 0 ? 'Carousel-hidden fas fa-chevron-circle-left fa-2x' : 'fas fa-chevron-circle-left fa-2x';
  return (
    <div className="Carousel">
      <h1>{props.title}</h1>
      <div className="Carousel-main">
        <i
          className={leftHide}
          onClick={goBackward}
          data-testid="left-arrow"
        />
        <Card
          caption={card.caption}
          src={card.src}
          currNum={cardIdx + 1}
          totalNum={total}
        />
        <i
          className={rightHide}
          onClick={goForward}
          data-testid="right-arrow"
        />
      </div>
    </div>
  );
}

Carousel.defaultProps = {
  cardData: [
    {
      src: image1,
      caption: "Photo by Richard Pasquarella on Unsplash"
    },
    {
      src: image2,
      caption: "Photo by Pratik Patel on Unsplash"
    },
    {
      src: image3,
      caption: "Photo by Josh Post on Unsplash"
    }
  ],
  title: "Shells from far away beaches."
};

export default Carousel;
