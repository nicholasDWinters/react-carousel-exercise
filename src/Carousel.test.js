import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";

it("works when you click on the right arrow", function () {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  // expect the first image to show, but not the second
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = queryByTestId("right-arrow");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).toBeInTheDocument();
});


it("works when you click on the left arrow", function () {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  // move forward in the carousel
  const rightArrow = queryByTestId("right-arrow");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();

  // move backward in the carousel
  const leftArrow = queryByTestId('left-arrow');
  fireEvent.click(leftArrow);

  // expect the first image to show, but not the second
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();
});

it('should hide the arrow buttons when necessary', function () {
  const { queryByTestId } = render(<Carousel />);
  const rightArrow = queryByTestId('right-arrow');
  const leftArrow = queryByTestId('left-arrow');
  expect(leftArrow).toHaveClass('Carousel-hidden');
  expect(rightArrow).not.toHaveClass('Carousel-hidden');
  fireEvent.click(rightArrow);
  fireEvent.click(rightArrow);
  expect(rightArrow).toHaveClass('Carousel-hidden');
  expect(leftArrow).not.toHaveClass('Carousel-hidden');
})


it('renders without crashing', function () {
  render(<Carousel />);
});

it('matches snapshot', function () {
  const { asFragment } = render(<Carousel />);
  expect(asFragment()).toMatchSnapshot();
});

it('matches snapshot', function () {
  const { asFragment, queryByTestId } = render(<Carousel />);
  const rightArrow = queryByTestId('right-arrow');
  fireEvent.click(rightArrow);
  expect(asFragment()).toMatchSnapshot();
});

it('matches snapshot', function () {
  const { asFragment, queryByTestId } = render(<Carousel />);
  const rightArrow = queryByTestId('right-arrow');
  fireEvent.click(rightArrow);
  fireEvent.click(rightArrow);
  expect(asFragment()).toMatchSnapshot();
});