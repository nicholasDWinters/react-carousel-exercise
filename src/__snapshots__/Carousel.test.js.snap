// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Shells from far away beaches.
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="Carousel-hidden fas fa-chevron-circle-left fa-2x"
        data-testid="left-arrow"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          Photo by Richard Pasquarella on Unsplash
        </h4>
        <img
          alt="Photo by Richard Pasquarella on Unsplash"
          class="Card-image"
          src="image1.jpg"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        class="fas fa-chevron-circle-right fa-2x"
        data-testid="right-arrow"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 2`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Shells from far away beaches.
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="fas fa-chevron-circle-left fa-2x"
        data-testid="left-arrow"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          Photo by Pratik Patel on Unsplash
        </h4>
        <img
          alt="Photo by Pratik Patel on Unsplash"
          class="Card-image"
          src="image2.jpg"
        />
        <small
          class="Card-small"
        >
          Image 2 of 3.
        </small>
      </div>
      <i
        class="fas fa-chevron-circle-right fa-2x"
        data-testid="right-arrow"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 3`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Shells from far away beaches.
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="fas fa-chevron-circle-left fa-2x"
        data-testid="left-arrow"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          Photo by Josh Post on Unsplash
        </h4>
        <img
          alt="Photo by Josh Post on Unsplash"
          class="Card-image"
          src="image3.jpg"
        />
        <small
          class="Card-small"
        >
          Image 3 of 3.
        </small>
      </div>
      <i
        class="Carousel-hidden fas fa-chevron-circle-right fa-2x"
        data-testid="right-arrow"
      />
    </div>
  </div>
</DocumentFragment>
`;
